src/Umbraco.Core/Models/EntityContainer.cs

```diff
@@ -1,23 +1,93 @@
+        };
+
+        /// <summary>
+        /// Initializes a new instance of an <see cref="EntityContainer"/> class.
+        /// </summary>
+        public EntityContainer(Guid containedObjectType)
+        {
+            if (ObjectTypeMap.ContainsKey(containedObjectType) == false)
+                throw new ArgumentException("Not a contained object type.", "containedObjectType");
+            _containedObjectType = containedObjectType;
+
+            ParentId = -1;
+            Path = "-1";
+            Level = 0;
+            SortOrder = 0;
         }
 
-        public EntityContainer(int id, int parentId, string name, int userId, string path)
+        /// <summary>
+        /// Initializes a new instance of an <see cref="EntityContainer"/> class.
+        /// </summary>
+        internal EntityContainer(int id, Guid uniqueId, int parentId, string path, int level, int sortOrder, Guid containedObjectType, string name, int userId)
+            : this(containedObjectType)
         {
             Id = id;
+            Key = uniqueId;
             ParentId = parentId;
             Name = name;
-            CreatorId = userId;
             Path = path;
+            Level = level;
+            SortOrder = sortOrder;
+            CreatorId = userId;
+        }
+
+        /// <summary>
+        /// Gets or sets the node object type of the contained objects.
+        /// </summary>
+        public Guid ContainedObjectType
+        {
+            get { return _containedObjectType; }
+        }
+
+        /// <summary>
+        /// Gets the node object type of the container objects.
+        /// </summary>
+        public Guid ContainerObjectType
+        {
+            get { return ObjectTypeMap[_containedObjectType]; }
+        }
+
+        /// <summary>
+        /// Gets the container object type corresponding to a contained object type.
+        /// </summary>
+        /// <param name="containedObjectType">The contained object type.</param>
+        /// <returns>The object type of containers containing objects of the contained object type.</returns>
+        public static Guid GetContainerObjectType(Guid containedObjectType)
+        {
+            if (ObjectTypeMap.ContainsKey(containedObjectType) == false)
+                throw new ArgumentException("Not a contained object type.", "containedObjectType");
+            return ObjectTypeMap[containedObjectType];
+        }
+
+        /// <summary>
+        /// Gets the contained object type corresponding to a container object type.
+        /// </summary>
+        /// <param name="containerObjectType">The container object type.</param>
+        /// <returns>The object type of objects that containers of the container object type can contain.</returns>
+        public static Guid GetContainedObjectType(Guid containerObjectType)
+        {
+            var contained = ObjectTypeMap.FirstOrDefault(x => x.Value == containerObjectType).Key;
```

Comment:
My bad, in fact this is right: ObjectTypeMap is a dictionary, FirstOrDefault will return the first KeyValuePair that matches, or the default KeyValuePair, which is _not_ null because KeyValuePair is a value type (struct), so it should return an actual KeyValuePair object with its .Key being null. Making sense?